### ISSUE TITLE
[Feature request: Completion] match from formatted reference

### ISSUE THREAD

**Issue by clason** [latex-lsp/texlab#27]
(Another one from https://github.com/lervag/vimtex; not sure if it's doable server-side. Vimtex parses the aux file for this, so requires a compilation run before it works; texlab seems smarter in that it does not require this? I have to admit that I'm not completely sure how the LSP completion actually works...)

Would it be possible to offer completion based on the actual *contents* of the reference? For example, in vimtex I can do the following (`<c-x c-o>` is the default keystroke to trigger completion):

1. `\eqref{2.1<c-x c-o>` and get (among others) `\eqref{eq:integral`, if the equation has the tag "(2.1)"; similarly for chapters, sections, theorems etc.
2. `\cite{Hilb<c-x c-o>` and get (among others) `\cite{Courant:1953`, if the bib entry is 
```
@book{Courant:1953,
  author = {Courant, Richard and Hilbert, David},
  title = {Methods of Mathematical Physics},
  year = {1953},
}
```
and similar with `\cite{Physics<c-x c-o>` , `\cite{1953<c-x c-o>`, etc.

**Comment by pfoerster** [comment]
With the latest commit of the master branch, TexLab should offer completion based on the contents of the reference. For labels, the contents include the section and the caption of the label. Note that TexLab is not parsing the aux file at the moment. That's why it is not possible to complete a label based on its number yet. However, we are looking to change this in the future. Note that this feature does not work with VSCode because it seems to ignore the `filterText` property (see https://github.com/Microsoft/vscode/issues/26874). I tested this feature with coq.nvim so I'm not sure if it will work with vim-lsp too.

**Comment by clason** [comment]
Nice! It doesn't seem to work with vim-lsp, unfortunately. If you can give me a short example (maybe for labels as well as citations) that should work and a brief explanation how this is realized using `filterText`, I could open an issue about it at vim-lsp with a bit more details.

**Comment by pfoerster** [comment]
The following example should work with `coc.nvim`:
```latex
\documentclass{article}
\begin{document}

\section*{Foo Bar Baz}%
\label{sec:foo}

\section*{Qux}
\label{sec:qux}

\ref{Bar|}

\end{document}
```
`|` is the position of the cursor and the completion list should show `sec:foo` as a possible result.
The completion item has the following (simplified) form
```json
{
   "label": "sec:foo",
   "filterText": "sec:foo Foo Bar Baz",
   ...
}
```
According to the LSP specification, `filterText` is used instead of the `label` to filter the set of completion items. Looking at the source code of `vim-lsp`, I can see that there is no usage of `filterText`.

**Comment by clason** [comment]
Thanks, reported at `vim-lsp`: https://github.com/prabirshrestha/vim-lsp/issues/438

(Regarding parsing `aux` files: That's fine; I think it's a plus that texlab doesn't require `aux` files and thus works even before compiling the document. Not matching references is a small price to pay; this feature is more useful for citations, anyway. I assume a citation completion item contains as `filterText` the bibkey, followed by the same formatted reference as for hover?)

**Comment by clason** [comment]
@pfoerster A followup: I noticed that the filter text is always the containing section name, even for theorems, equations and so on. That's not so useful for filtering :) If there is no formatted reference (yet) for theorems, equations, etc., the default `filterText` should just contain the label. (For theorems and other environments, there's sometimes a name provided using `\begin{theorem}[Banach--Steinhaus]` -- if it's feasible to extract that, this would also be a good default.)

The same thing happens with `hover`, by the way: *Any* reference gets the containing section name as the hover text. Should I open a separate issue for this?

**Comment by pfoerster** [comment]
@clason Can you try if the latest release of TexLab works for you? In your example, the hover text should now be `Theorem (Banach--Steinhaus)` and the `filterText` should contain the name of the theorem.

**Comment by clason** [comment]
@pfoerster Thank you! It does work, but in multi-file projects only within a single file (otherwise I get "No hover information found"); I'll try to get a log. The filter text looks correct, but I notice that `equation` environments and similar still have the surrounding section name as `hover` and `filterText`?

**Comment by clason** [comment]
Weird, on another (theorem) label, I get an image preview...

**Comment by clason** [comment]
I think the problem is that more complicated documents with many definitions break the hover on equations (even in vs code, I get a messed up image), which also breaks the hover on labels. Is there an easy way to disable the equation hover (I'm building from source) to check?

**Comment by clason** [comment]
I think there's a problem opening other files from the project for obtaining hover previews. In the `vim-lsp` log, I see a long string of 
```
Wed 21 Aug 2019 05:04:27 PM CEST:["<---", 5, "texlab", {"response": {"method":             "textDocument/publishDiagnostics", "jsonrpc": "2.0", "params": {"diagnostics": [], "uri":  "file://XXXX"}}}]  
Wed 21 Aug 2019 05:04:27 PM CEST:["<---", 5, "texlab", {"response": {"method":             "textDocument/publishDiagnostics", "jsonrpc": "2.0", "params": {"diagnostics": [], "uri":  "file://YYYY"}}}]                          
Wed 21 Aug 2019 05:04:27 PM CEST:["s:on_stdout on_notification() error", "Vim(call):E5555: API call: Wrong type for argument 1, expecting Buffer"]                                    
```
where `XXXX` is the current file and `YYYY` are (correctly identified) other project files. 

I have the same issues with lack of or incorrect hover in VS Code, but can't debug that.

**Comment by clason** [comment]
Here's a minimal reproducing example: Hovering on the `\ref` in `test2.tex` doesn't show anything (in vim or VS Code). The containing files are correctly recognized, though, because autocomplete works.
[test.zip](https://github.com/latex-lsp/texlab/files/3526004/test.zip)

**Comment by pfoerster** [comment]
> Here's a minimal reproducing example: Hovering on the \ref in test2.tex doesn't show anything (in vim or VS Code). The containing files are correctly recognized, though, because autocomplete and peek definition work.

The example you provided is missing a \newtheorem in order to work.
Nevertheless, I managed to fix a bug that may be causing the issue.
In addition to that, I increased the range when hovering over labels (before this change you had to hover over the name in order to trigger the tooltip, now you can hover over the command itself).
This should prevent the preview from showing up when it shouldn't.

> Is there an easy way to disable the equation hover (I'm building from source) to check?

You can disable the preview feature by commenting out line 31 in src/hover/mod.rs.

>  The filter text looks correct, but I notice that equation environments and similar still have the surrounding section name as hover and `filterText`

What kind of filter text are you expecting when hovering over an equation? Just the label name and the word "Equation"?

**Comment by clason** [comment]
> The example you provided is missing a \newtheorem in order to work.

True, but I got the same behavior if the `\newtheorem` is there -- sorry for it not really being a reproducing example. (I did not realize texlab was that smart in parsing...) The fix looks like it helps, thank you very much! For the previously missing hover labels, the position of the popup is still off (too far to the right), but I assume that's due to `vim-lsp`, not `texlab`. 

> This should prevent the preview from showing up when it shouldn't.

Ah, so that's what was happening. Now it makes sense.

> What kind of filter text are you expecting when hovering over an equation? Just the label name and the word "Equation"?

Sorry, I wasn't quite clear on this. I'm only talking about labels -- hovering or completing. (The math preview is cool, but useless for me since I use a text-based editor that can't parse markdown.)

Not sure what would make sense for hovering over an equation label -- if in doubt, I'd say show nothing (or as you propose, "Equation"), but I can see the rationale for showing the containing section. If somebody else finds it useful, that's sufficient argument for me!

For filter text, I similarly would have expected something specific to _that_ reference -- if it matches many (possibly _many_) labels, it doesn't seem so useful to me. So again, I'd have suggested just sticking to the actual label for the filterText (or omitting filterText, which should have the same effect?). But I forgot that `\eqref` only matches math now, so completing `\eqref{<section-name>` would show you all equations (and only equations) from that section, and that might conceivably be useful.

So if that was actually intended behavior, feel free to ignore that comment!

**Comment by clason** [comment]
Just noticed a minor thing: The sectioning commands for which hover works don't seem to include `\part`?

(Also, I just saw that for figures, hover shows the caption -- that's really neat! 💯 )

**Comment by pfoerster** [comment]
> \part is indeed missing from language.json (it has level -1).

`\part` should now work as intended.

**Comment by clason** [comment]
So it does, thank you very much!

**Comment by clason** [comment]
Sorry to keep coming back to this, but here's another (minor) thing I just noticed: The `enumerate` environment is missing -- I just get the name of the containing section here. It would be nice to have this also parsed as `item` plus the label (that is put into the aux file. Here's a minimal example:
```tex
\documentclass{article}
\begin{document}
\begin{enumerate}
    \item \label{it:first} one
    \item \label{it:second} two
    \item \label{it:third} three
\end{enumerate}
\end{document}
```
and the corresponding `aux` file
```
\relax 
\newlabel{it:first}{{1}{1}}
\newlabel{it:second}{{2}{1}}
\newlabel{it:third}{{3}{1}}
```

There's a bit of a hitch, though -- if you include the `enumitem` package, the `aux` file is different:
```tex
\documentclass{article}
\usepackage{enumitem}
\begin{document}
\begin{enumerate}[label=(\roman*)]
    \item \label{it:first} one
    \item \label{it:second} two
    \item \label{it:third} three
\end{enumerate}
\end{document}
```
```
\relax 
\newlabel{it:first}{{{{(i)}}}{1}}
\newlabel{it:second}{{{{(ii)}}}{1}}
\newlabel{it:third}{{{{(iii)}}}{1}}
```
And finally, if you do the same manually, you get
```tex
\documentclass{article}
\begin{document}
\begin{enumerate}
    \item[(i)] \label{it:first} one
    \item[(ii)] \label{it:second} two
    \item[(iii)] \label{it:third} three
\end{enumerate}
\end{document}
```
```
\relax 
\newlabel{it:first}{{}{1}}
\newlabel{it:second}{{}{1}}
\newlabel{it:third}{{}{1}}
```
(but I think that can be taken care of by the `[...]` parsing as in theorems.)

(There's also the `enumerate` package, though...)

In any case, having this parsed as `item` instead of `<section name>` would be useful even without the full label.

**Comment by clason** [comment]
Oh, and for `enumerate`, it would also be nice to have `locationLink` support -- or rather, the `targetRange` for `item` labels should be the surrounding `enumerate` environment, if possible. (Or would that be confusing?)

**Comment by clason** [comment]
And also add support for Document/WorkspaceSymbol (announce as `item` instead of `enumerate`)?

**Comment by pfoerster** [comment]
> Sorry to keep coming back to this, but here's another (minor) thing I just noticed: The enumerate environment is missing -- I just get the name of the containing section here. It would be nice to have this also parsed as item plus the label (that is put into the aux file. Here's a minimal example:

This should be implemented now with the latest commit on the master branch. Thank you for the examples you provided. The only thing that is missing at the moment is support for the `enumerate` package.

> Oh, and for enumerate, it would also be nice to have locationLink support -- or rather, the targetRange for item labels should be the surrounding enumerate environment, if possible. (Or would that be confusing?)

Actually, I'm not sure about this one. Another possibility would be to use the item's range. At the moment, the master branch uses the range of the `enumerate` environment.

> And also add support for Document/WorkspaceSymbol (announce as item instead of enumerate)?

The list of items should appear in the result list.

**Comment by clason** [comment]
Thank you! I tested it, and it works nicely, with one small issue: For something like
```latex
\begin{enumerate}
\item ....
\begin{equation}\label{eq1}
...
\end{equation}
\end{enumerate}
```
the symbol shows the equation number rather than `item`, while the equation itself does not show up in the list of symbols. If I add a label to the item, it's shown correctly, but the equation is still lost.

**Comment by clason** [comment]
> The only thing that is missing at the moment is support for the enumerate package.

Is there some fundamental difficulty, or just lack of examples (which I'd be happy to provide) or lack of time (which is completely understandable)?

> Actually, I'm not sure about this one. Another possibility would be to use the item's range. At the moment, the master branch uses the range of the enumerate environment.

I'm not sure, either. The current behavior is certainly fine, but you're right that the `\item` range would be more consistent (if easily implemented).

**Comment by pfoerster** [comment]
> Is there some fundamental difficulty, or just lack of examples (which I'd be happy to provide) or lack of time (which is completely understandable)?

It is mostly a lack of time and rather difficult to implement. Basically you have to implement the logic of the enumerate package yourself.

>  I'm not sure, either. The current behavior is certainly fine, but you're right that the \item range would be more consistent (if easily implemented).

I need to calculate the item range anyway, so there is no difference in implementation complexity.

> Thank you! I tested it, and it works nicely, with one small issue: For something like
\begin{enumerate}
\item ....
\begin{equation}\label{eq1}
...
\end{equation}
\end{enumerate}
the symbol shows the equation number rather than item, while the equation itself does not show up in the list of symbols. If I add a label to the item, it's shown correctly, but the equation is still lost.

Thanks for the report. I will look into that one.

**Comment by clason** [comment]
For the sake of completeness (and while I'm at it), here's the above example with `enumerate`:
```tex
\documentclass{article}
\usepackage{enumerate}
\begin{document}
\begin{enumerate}[(i)]
    \item \label{it:first} one
    \item \label{it:second} two
    \item \label{it:third} three
\end{enumerate}
\end{document}
```
with `aux` file
```
\relax
\newlabel{it:first}{{i}{1}}
\newlabel{it:second}{{ii}{1}}
\newlabel{it:third}{{iii}{1}}
```
It actually already works out of the box; it's just missing the parentheses around the labels. That's a really minor issue.

**Comment by clason** [comment]
A related question (maybe better posed as a comment to the other issue): how are the symbols sorted? At least vim-lsp shows them in some sort of reverse order (e.g., the containing `enumerate` after the `items`, the `section` after all symbols in that section, etc.)

**Comment by pfoerster** [comment]
> A related question (maybe better posed as a comment to the other issue): how are the symbols sorted? At least vim-lsp shows them in some sort of reverse order (e.g., the containing enumerate after the items, the section after all symbols in that section, etc.) And enumerate and equations are sorted before all others. (It's probably a problem with vim-lsp that doesn't understand hierarchical DocumentSymbols, though.)

At the moment, there are not sorted at all. The order roughly correlates with the appearance in the syntax tree (leaves appear first). I think sorting them by their appearance in the text is probably a good idea.

**Comment by clason** [comment]
> I think sorting them by their appearance in the text is probably a good idea.

That would of course be nice. I think swapping the order of leaves and root in each recursion would already improve things (if that's easy).

**Comment by pfoerster** [comment]
@clason Can you try 033d93f4d1fcd2adf173300efab5e253ca74cbc7?

**Comment by clason** [comment]
Fantastic! (Now I'm getting greedy: `workspaceSymbols` are sorted first alphabetically by filename -- can that also be sorted based on the project tree, or is that additional work?)

Regarding the above issue with equations in items: Now both item and equation show up in the list -- all that's left is the minor cosmetic issue that it show `enum member : 8.42` instead of `enum member: Item`.

**Comment by pfoerster** [comment]
> Regarding the above issue with equations in items: Now both item and equation show up in the list -- all that's left is the minor cosmetic issue that it show enum member : 8.42 instead of enum member: Item. (I guess that was the case before as well; I just didn't realize that enums and equations were sorted and shown separately.)

This one should be fixed with a4c10d14d2a57bbe9b732c2da8f0613d5379ebf1.

> Fantastic! (Now I'm getting greedy: workspaceSymbols are sorted first alphabetically by filename -- can that also be sorted based on the project tree, or is that additional work? If it is, the current behavior is fine, too.)

What exactly do you mean by getting sorted based on the project tree? Can you give an example?

**Comment by clason** [comment]
I can create an actual example in a moment, but if I have `main.tex` that includes -- in this order -- `b.tex` and `a.tex`, a `workspaceSymbols` request gives the (correctly sorted) symbols in `a.tex` before those in `b.tex` before those in `main.tex`, while the other way around would be more natural.

(Might be due to the client, though.)

**Comment by pfoerster** [comment]
> I can create an actual example in a moment, but if I have main.tex that includes -- in this order -- b.tex and a.tex, a workspaceSymbols request gives the (correctly sorted) symbols in a.tex before those in b.tex before those in main.tex, while the other way around would be more natural.

Ah okay, now I understand. This one should be possible to implement with a bit of extra work.

**Comment by clason** [comment]
> This one should be fixed with a4c10d1.

You implement bug fixes and feature requests in less time than it takes me to compile the server and test it ;) Works nicely!

> Ah okay, now I understand. This one should be possible to implement with a bit of extra work.

Not a big priority; vim-lsp prefixes the `workspaceSymbol` with the filename, so this is not actually confusing.

**Comment by clason** [comment]
Ah, and I see the issue with the `enumerate` package: you only see the counter (alphanumeric or roman) in the `aux` file, not the static styling. For me this isn't really worth putting additional work in -- feel free to close this issue, and thanks for implementing this!

(I'll copy the comment about sorting `workspaceSymbols` to the other issue.)
